src/dali_executor/CMakeLists.txt

```diff
@@ -30,25 +30,27 @@ include(${tritondalibackend_SOURCE_DIR}/cmake/dali.cmake)
 
 separate_arguments(EXTRA_DOWNLOAD_ARGS UNIX_COMMAND ${DALI_DOWNLOAD_EXTRA_OPTIONS})
 
-add_custom_command(  # Download and unpack DALI wheel
-    OUTPUT dali_whl
-    COMMAND
-        pip download ${EXTRA_DOWNLOAD_ARGS} --extra-index-url ${DALI_EXTRA_INDEX_URL}
-          -d ${CMAKE_CURRENT_BINARY_DIR}
-          ${DALI_DOWNLOAD_PKG_NAME}$<$<BOOL:${DALI_VERSION}>:==${DALI_VERSION}>
+if(EXTRA_DOWNLOAD_ARGS)
+    configure_file(${tritondalibackend_SOURCE_DIR}/cmake/dalienv-extra_args.yml.in dalienv.yml @ONLY)
+else()
+    configure_file(${tritondalibackend_SOURCE_DIR}/cmake/dalienv.yml.in dalienv.yml @ONLY)
+endif()
+
+add_custom_command(
+    OUTPUT dali_env
     COMMAND
-        unzip -q -o ${CMAKE_CURRENT_BINARY_DIR}/nvidia_dali*.whl -d ${CMAKE_CURRENT_BINARY_DIR}/dali
-    COMMAND touch dali_whl  # So that this command won't be re-run every time
-    COMMENT "Acquiring DALI release"
+        conda env create -f dalienv.yml -p ${CMAKE_CURRENT_BINARY_DIR}/dalienv -q
+    COMMAND touch dali_env  # So that this command won't be re-run every time
+    COMMENT "Building DALI conda environment."
 )
 
-if (${TRITON_DALI_SKIP_DOWNLOAD})
+if(${TRITON_DALI_SKIP_DOWNLOAD})
     get_dali_paths(DALI_INCLUDE_DIR DALI_LIB_DIR DALI_LIBRARIES)
-else ()
-    set(DALI_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/dali/nvidia/dali/include)
-    set(DALI_LIB_DIR ${CMAKE_CURRENT_BINARY_DIR}/dali/nvidia/dali)
+else()
+    set(DALI_INCLUDE_DIR ${CMAKE_CURRENT_BINARY_DIR}/dalienv/lib/python3.8/site-packages/nvidia/dali/include)
```

Comment:
Maybe we should not hardcode it but something like:
```
conda run -n dalienv python -c "import nvidia.dali.sysconfig; print(nvidia.dali.sysconfig.get_include_dir()"
```
and
```
conda run -n dalienv python -c "import nvidia.dali.sysconfig; print(nvidia.dali.sysconfig.get_lib_dir()"
```

Reply:
Good point. I'll change this.

Reply:
So maybe you should use `execute_process` instead of `add_custom_command` ?
Is there any value in making the creation of the conda env a build step instead of part of the configuration process?

Reply:
The value is that it's conducted once per target build. In a hypothetical scenario, that somebody goes:
```
cmake -D TRITON_DALI_SKIP_DOWNLOAD=OFF ..
make
```
And then, during the development touches any of the `CMakeLists.txt` files, consecutive `make` command won't re-download and re-create the conda env. Which is forbidden - user would need to remove entire `build` directory and run `cmake` from scratch.

Reply:
Tried this one. This also has hard time working, because of `configure_file` command above: we need the configured file to create conda environment. OTOH, `configure_file` happens at build stage, with all its repercussions, e.g. if the target is not linked to any executable/library, the file won't be generated. This makes the commands that rely on `dalienv.yml` no-op in the generation step (because this file doesn't exist yet).

Judging all the tweaks here, I believe it is better to hardcode the library paths. It in an easy error to spot - the build will break if there are no libraries at that location, and an easy error to fix if necessary - if the library paths will change, we just need to hardcode different path.

Also, this is not really a technical point, but my gut (and slight experience with CMake) tells me that the build stage is better suitable for workloads such as downloading dependencies. An argument for this might be, that it happens concurrently - when we go `make -j`, one thread handles the download, while everything else that does not depend on the downloading target might be built simultaneously.

Taking into account all this, I believe it would be better to keep creating the conda environment in the build stage, which (unfortunately) stipulates hardcoding the library paths. But, as said earlier, resolving error that might occur here would be rather easy.

Reply:
Up to you.

> Also, this is not really a technical point, but my gut (and slight experience with CMake) tells me that the build stage is better suitable for workloads such as downloading dependencies. An argument for this might be, that it happens concurrently - when we go make -j, one thread handles the download, while everything else that does not depend on the downloading target might be built simultaneously.

To make sure that it works as intended you need to add a dependency between the DALI backed binaries and `dali_env`. Otherwise, it will attempt to use includes from the env that is just created in parallel.

Reply:
..., ok, I see L63. NVM then.